src/mem3/src/mem3_reshard.erl

```diff
@@ -0,0 +1,919 @@
+    TrimmedRev = lists:sublist(UpdatedRev, max_history()),
+    lists:reverse(TrimmedRev).
+
+
+-spec shard_from_name(binary()) -> #shard{}.
+shard_from_name(<<"shards/", _:8/binary, "-", _:8/binary, "/",
+        Rest/binary>> = Shard) ->
+    Range = mem3:range(Shard),
+    [DbName, Suffix] = binary:split(Rest, <<".">>),
+    build_shard(Range, DbName, Suffix).
+
+
+% For debugging only
+
+-spec reset_state() -> ok.
+reset_state() ->
+    gen_server:call(?MODULE, reset_state, infinity).
+
+
+% Gen server functions
+
+init(_) ->
+    % Advertise resharding API feature only if it is not disabled
+    case is_disabled() of
+        true -> ok;
+        false -> config:enable_feature('reshard')
+    end,
+    couch_log:notice("~p start init()", [?MODULE]),
+    EtsOpts = [named_table, {keypos, #job.id}, {read_concurrency, true}],
+    ?MODULE = ets:new(?MODULE, EtsOpts),
+    ManagerPid = self(),
+    State = #state{
+        state = running,
+        state_info = [],
+        update_time = now_sec(),
+        node = node(),
+        db_monitor = spawn_link(fun() -> db_monitor(ManagerPid) end)
+    },
+    State1 = mem3_reshard_store:init(State, ?JOB_PREFIX, state_id()),
+    State2 = mem3_reshard_store:load_state(State1, running),
+    State3 = maybe_disable(State2),
+    gen_server:cast(self(), reload_jobs),
+    {ok, State3}.
+
+
+terminate(Reason, State) ->
+    couch_log:notice("~p terminate ~p ~p", [?MODULE, Reason, statefmt(State)]),
+    catch unlink(State#state.db_monitor),
+    catch exit(State#state.db_monitor, kill),
+    [kill_job_int(Job) || Job <- running_jobs()],
+    ok.
+
+
+handle_call(start, _From, #state{state = stopped} = State) ->
+    State1 = State#state{
+        state = running,
+        update_time = now_sec(),
+        state_info = info_delete(reason, State#state.state_info)
+    },
+    ok = mem3_reshard_store:store_state(State1),
+    State2 = maybe_disable(State1),
+    State3 = reload_jobs(State2),
+    {reply, ok, State3};
+
+handle_call(start, _From, State) ->
+    {reply, ok, State};
+
+handle_call({stop, Reason}, _From, #state{state = running} = State) ->
+    State1 = State#state{
+        state = stopped,
+        update_time = now_sec(),
+        state_info = info_update(reason, Reason, State#state.state_info)
+    },
+    ok = mem3_reshard_store:store_state(State1),
+    [temporarily_stop_job(Job) || Job <- running_jobs()],
+    {reply, ok, State1};
+
+handle_call({stop, _}, _From, State) ->
+    {reply, ok, State};
+
+handle_call({start_job, #job{id = Id, source = Source} = Job}, _From, State) ->
+    couch_log:notice("~p start_job call ~p", [?MODULE, jobfmt(Job)]),
+    Total = ets:info(?MODULE, size),
+    SourceOk = mem3_reshard_validate:source(Source),
+    case {job_by_id(Id), Total + 1 =<  get_max_jobs(), SourceOk} of
+        {not_found, true, ok} ->
+            handle_start_job(Job, State);
+        {#job{}, _, _} ->
+            {reply, {error, job_already_exists}, State};
```

Comment:
I used `error` thinking it would be more explicit. It would force the user to wait for the job to complete examine the result then remove it and re-create. Otherwise if they add the same job twice they could think they might have split the db twice, but in fact on second call they just got the same ID back from the already running job?